feat: Disable CWIP Accounting checkbox added in Company and Asset Category

marination opened the pull request:
- **Asset Settings** is removed completely
- **'Enable Capital Work in Progress Accounting'** checkbox will be present in **Company** under **Fixed Asset Depreciation Settings**

![Screenshot 2019-10-11 at 3 38 34 PM](https://user-images.githubusercontent.com/25857446/66643840-34880380-ec3d-11e9-92dc-fd3d0dcae411.png)

- Also in **Asset Category** under **Depreciation Options**

![Screenshot 2019-10-11 at 3 37 06 PM](https://user-images.githubusercontent.com/25857446/66643914-5c776700-ec3d-11e9-9f6c-57c2bcd2ebba.png)

- It will be globally enabled if enabled in **Company** , else check will be on **Asset Category** level

![Screenshot 2019-10-11 at 3 42 43 PM](https://user-images.githubusercontent.com/25857446/66644130-cee84700-ec3d-11e9-80a1-06199c2898e8.png)

- Removed some non-functional code

marination committed:
feat: Disable CWIP Accounting checkbox added in Company and Asset Category

Asset Settings is removed completely
Disable CWIP Accounting checkbox will give priority to Asset Category

nabinhait commented on erpnext/accounts/doctype/purchase_invoice/purchase_invoice.py:
```diff
@@ -225,6 +225,7 @@ def set_expense_account(self, for_validate=False):
 			# in case of auto inventory accounting,
 			# expense account is always "Stock Received But Not Billed" for a stock item
 			# except epening entry, drop-ship entry and fixed asset items
+			asset_category = frappe.db.get_value("Item", item.item_code, "asset_category")
```
get_cached_value

nabinhait commented on erpnext/accounts/general_ledger.py:
```diff
@@ -148,14 +148,17 @@ def validate_account_for_perpetual_inventory(gl_map):
-	if not cint(frappe.db.get_value("Asset Settings", None, "disable_cwip_accounting")) \
+	from erpnext.assets.doctype.asset.asset import is_cwip_accounting_disabled
+	
+	if not is_cwip_accounting_disabled(gl_map[0].company, None) \
```
Check if CWIP is enabled for any of the asset category

nabinhait reviewed (changes requested):
- Rename `disable cwip accounting` to `Enable Capital Work in Progress Accounting`
- Make "Enable Capital Work in Progress Accounting" a Select field with options "Yes", "No" and "", default is "No". Patch for existing accounts

marination committed:
fix: Changed checkbox name to 'Enable Capital Work in Progress Accounting'

- checkbox will be disabled by default
- Enabling it in Company will globally enable it
- When globally disabled , it's value on the asset category will be considered

marination committed:
chore: Added patch to set pre-existing CWIP checkbox value into new checkbox

Anurag810 committed:
fix: Asset Test and Patch

Anurag810 committed:
fix(test): Asset

Anurag810 committed:
Merge branch 'asset-checkbox' of https://github.com/marination/erpnext into asset-checkbox

Anurag810 committed:
fix(test): Opening Invoice Creation Tool

marination committed:
Merge branch 'develop' into asset-checkbox

nabinhait committed:
Update asset.py

nabinhait commented on erpnext/patches/v12_0/set_cwip_and_delete_asset_settings.py:
```diff
@@ -0,0 +1,27 @@
+			and field='disable_cwip_accounting' """, as_dict=1)
+		company = frappe.db.get_single_value("Global Defaults", "default_company")
+
+		if company:
```
Always set enable_cwip_accounting value for all companies

nabinhait commented on erpnext/patches/v12_0/set_cwip_and_delete_asset_settings.py:
```diff
@@ -0,0 +1,27 @@
+			for company in companies:
+				company_doc = frappe.get_doc("Company", company)
+				company_doc.enable_cwip_accounting = cint(not cint(cwip_value[0]['value']))
+				company_doc.save()
```
Use `set_value` instead of doing `get_doc` and `save`

nabinhait commented on erpnext/accounts/doctype/purchase_invoice/purchase_invoice.py:
```diff
@@ -404,6 +408,15 @@ def get_gl_entries(self, warehouse_account=None):
+	def check_asset_cwip_enabled(self):
+		# Check if there exists any item with cwip accounting enabled in it's asset category
+		for item in self.get("items"):
+			if item.item_code:
```
```
if item.item_code and item.is_fixed_asset:
	asset_category = frappe.get_cached_value("Item", item.item_code, "asset_category")
	if is_cwip_accounting_enabled(self.company, asset_category):
		return 1
```

nabinhait commented on erpnext/accounts/doctype/purchase_invoice/purchase_invoice.py:
```diff
@@ -528,7 +544,10 @@ def make_item_gl_entries(self, gl_entries):
 	def get_asset_gl_entry(self, gl_entries):
 		for item in self.get("items"):
-			if item.is_fixed_asset:
+			if item.item_code:
```
`if item.item_code and item.is_fixed_asset:`

nabinhait commented on erpnext/accounts/general_ledger.py:
```diff
@@ -148,14 +148,28 @@ def validate_account_for_perpetual_inventory(gl_map):
-	if not cint(frappe.db.get_value("Asset Settings", None, "disable_cwip_accounting")) \
+	from erpnext.assets.doctype.asset.asset import is_cwip_accounting_enabled
+	asset_category = None
+	enable_cwip_in_company = cint(frappe.get_cached_value("Company",
```
The code should look like this:

```
cwip_enabled = cint(frappe.get_cached_value("Company", gl_map[0].company, "enable_cwip_accounting"))

if not cwip_enabled:
	cwip_enabled = any([cint(ac.enable_cwip_accounting)
		for ac in frappe.db.get_all("Asset Category", "enable_cwip_accounting")])

if cwip_enabled and gl_map[0].voucher_type == "Journal Entry":
```

marination committed:
fix: Patch and other fixes

nabinhait committed:
Merge branch 'develop' into asset-checkbox

MorezMartin commented:
Hello
Little problem with it when trying to migrate 
<pre>Traceback (most recent call last):
  File &quot;/usr/lib/python3.5/runpy.py&quot;, line 193, in _run_module_as_main
    &quot;__main__&quot;, mod_spec)
  File &quot;/usr/lib/python3.5/runpy.py&quot;, line 85, in _run_code
    exec(code, run_globals)
  File &quot;/home/frappe/frappe-bench/apps/frappe/frappe/utils/bench_helper.py&quot;, line 97, in &lt;module&gt;
    main()
  File &quot;/home/frappe/frappe-bench/apps/frappe/frappe/utils/bench_helper.py&quot;, line 18, in main
    click.Group(commands=commands)(prog_name=&apos;bench&apos;)
  File &quot;/home/frappe/frappe-bench/env/lib/python3.5/site-packages/click/core.py&quot;, line 764, in __call__
    return self.main(*args, **kwargs)
  File &quot;/home/frappe/frappe-bench/env/lib/python3.5/site-packages/click/core.py&quot;, line 717, in main
    rv = self.invoke(ctx)
  File &quot;/home/frappe/frappe-bench/env/lib/python3.5/site-packages/click/core.py&quot;, line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File &quot;/home/frappe/frappe-bench/env/lib/python3.5/site-packages/click/core.py&quot;, line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File &quot;/home/frappe/frappe-bench/env/lib/python3.5/site-packages/click/core.py&quot;, line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File &quot;/home/frappe/frappe-bench/env/lib/python3.5/site-packages/click/core.py&quot;, line 555, in invoke
    return callback(*args, **kwargs)
  File &quot;/home/frappe/frappe-bench/env/lib/python3.5/site-packages/click/decorators.py&quot;, line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File &quot;/home/frappe/frappe-bench/apps/frappe/frappe/commands/__init__.py&quot;, line 25, in _func
    ret = f(frappe._dict(ctx.obj), *args, **kwargs)
  File &quot;/home/frappe/frappe-bench/apps/frappe/frappe/commands/site.py&quot;, line 233, in migrate
    migrate(context.verbose, rebuild_website=rebuild_website, skip_failing=skip_failing)
  File &quot;/home/frappe/frappe-bench/apps/frappe/frappe/migrate.py&quot;, line 48, in migrate
    frappe.modules.patch_handler.run_all(skip_failing)
  File &quot;/home/frappe/frappe-bench/apps/frappe/frappe/modules/patch_handler.py&quot;, line 41, in run_all
    run_patch(patch)
  File &quot;/home/frappe/frappe-bench/apps/frappe/frappe/modules/patch_handler.py&quot;, line 30, in run_patch
    if not run_single(patchmodule = patch):
  File &quot;/home/frappe/frappe-bench/apps/frappe/frappe/modules/patch_handler.py&quot;, line 71, in run_single
    return execute_patch(patchmodule, method, methodargs)
  File &quot;/home/frappe/frappe-bench/apps/frappe/frappe/modules/patch_handler.py&quot;, line 91, in execute_patch
    frappe.get_attr(patchmodule.split()[0] + &quot;.execute&quot;)()
  File &quot;/home/frappe/frappe-bench/apps/erpnext/erpnext/patches/v12_0/set_cwip_and_delete_asset_settings.py&quot;, line 17, in execute
    enable_cwip_accounting = cint(not cint(cwip_value[0][&apos;value&apos;]))
IndexError: list index out of range
</pre>

marination commented:
@MorezMartin fixed https://github.com/frappe/erpnext/pull/19576

MorezMartin commented:
Thanks a lot, i disabled the patch
